wgengine/wgtun/tun.go

```diff
@@ -0,0 +1,249 @@
+	logf logger.Logf
+	// tdev is the underlying TUN device.
+	tdev tun.Device
+
+	// buffer stores the oldest unconsumed packet from tdev.
+	// It is made a static buffer in order to avoid graticious allocation.
+	buffer [readMaxSize]byte
+	// bufferConsumed synchronizes access to buffer (shared by Read and poll).
+	bufferConsumed chan struct{}
+
+	// closed signals poll (by closing) when the device is closed.
+	closed chan struct{}
+	// errors is the error queue populated by poll.
+	errors chan error
+	// outbound is the queue by which packets leave the TUN device.
+	// The directions are relative to the network, not the device:
+	// inbound packets arrive via UDP and are written into the TUN device;
+	// outbound packets are read from the TUN device and sent out via UDP.
+	// This queue is needed because although inbound writes are synchronous,
+	// the other direction must wait on a Wireguard goroutine to poll it.
+	outbound chan []byte
+
+	// filterMu locks the filters, since they may be in use
+	// during the netmap refresh when ipn/local updates them.
+	filterMu    sync.Mutex
+	filterFlags filter.RunFlags
+	filter      *filter.Filter
+}
+
+func WrapTUN(logf logger.Logf, tdev tun.Device) *Device {
+	device := &Device{
+		logf:           logf,
+		tdev:           tdev,
+		bufferConsumed: make(chan struct{}),
+		closed:         make(chan struct{}),
+		errors:         make(chan error),
+		outbound:       make(chan []byte),
+		filterFlags:    filter.LogAccepts | filter.LogDrops,
+	}
+	go device.poll()
+	// The buffer starts out consumed.
+	device.bufferConsumed <- struct{}{}
+
+	return device
+}
+
+func (d *Device) Close() error {
+	// Other channels need not be closed: poll will exit gracefully after this.
+	close(d.closed)
+	return d.tdev.Close()
+}
+
+func (d *Device) Events() chan tun.Event {
+	return d.tdev.Events()
+}
+
+func (d *Device) File() *os.File {
+	return d.tdev.File()
+}
+
+func (d *Device) Flush() error {
+	return d.tdev.Flush()
+}
+
+func (d *Device) MTU() (int, error) {
+	return d.tdev.MTU()
+}
+
+func (d *Device) Name() (string, error) {
+	return d.tdev.Name()
+}
+
+// poll polls d.tdev.Read, placing the oldest unconsumed packet into d.buffer.
+// This is needed because d.tdev.Read in general may block (it does on Windows),
+// so packets may be stuck in d.outbound if d.Read called d.tdev.Read directly.
+func (d *Device) poll() {
+	for {
+		select {
+		case <-d.closed:
+			return
+		case <-d.bufferConsumed:
+			// continue
+		}
+
+		// Read may use memory in d.buffer before readOffset for mandatory headers.
+		// This is the rationale behind the tun.Device.{Read,Write} interfaces
+		// and the reason d.buffer has size MaxMessageSize and not MaxContentSize.
+		n, err := d.tdev.Read(d.buffer[:], readOffset)
+		if err != nil {
+			d.errors <- err
+			// In principle, read errors are not fatal (but wiregaurd-go disagrees).
+			d.bufferConsumed <- struct{}{}
+		} else {
+			d.outbound <- d.buffer[readOffset : readOffset+n]
```

Comment:
Consider:

- this goroutine makes it through the select above, d is not yet closed
- gets to here
- close(d.closed) is called elsewhere
- the other goroutine calling Read selects on <-d.closed and stops calling Read, so d.outbound is never read from
- this goroutine spends eternity trying to send on outbound.

I think you need to send with a select here, and put a read from d.closed in that select.

Reply:
Good point, thanks. I have also changed `bufferConsumed` to be buffered to avoid the scenario where `closed` is closed during `copy` in `Read`, and to make the error path here actually work.